docs/envvars.rst
Environment Variables
=====================

Normally, no environment variables need to be set. Most of the
environment variables used by Mesa/Gallium are for debugging purposes,
but they can sometimes be useful for debugging end-user issues.

Android System Properties
-------------------------

Android (generally) uses system properties rather than environment variables to
control Mesa/Gallium behavior, although there are some exceptions to this. See
:ref:`Android System Properties <android-android-system-properties>` for
details on naming and how to set and get system property values.

LibGL environment variables
---------------------------

.. envvar:: LIBGL_DEBUG

   If defined debug information will be printed to stderr. If set to
   ``verbose`` additional information will be printed.

.. envvar:: LIBGL_ALWAYS_INDIRECT

   if set to ``true``, forces an indirect rendering context/connection.

.. envvar:: LIBGL_ALWAYS_SOFTWARE

   if set to ``true``, always use software rendering

.. envvar:: LIBGL_NO_DRAWARRAYS

   if set to ``true``, do not use DrawArrays GLX protocol (for
   debugging)

.. envvar:: LIBGL_KOPPER_DISABLE

   disable Vulkan swapchains with Zink if set to ``true``.
   In general, this should not be used unless you know what you are
   doing. Some examples of "knowing what you are doing" include:
   - using a VK driver which has no WSI implementation for your display server
   - profiling the DRI frontend against your VK driver's WSI implementation

.. envvar:: LIBGL_KOPPER_DRI2

   if set to ``true``, allow loading Zink even if X11 does not support
   explicit DRM format modifiers. In general, this should not be used
   unless you know what you are doing. Some examples of "knowing what you
   are doing" include:
   - running xrdp
   - using a VK driver which doesn't support modifiers

Core Mesa environment variables
-------------------------------

.. envvar:: MESA_NO_ERROR

   if set to 1, error checking is disabled as per :ext:`GL_KHR_no_error`.
   This will result in undefined behavior for invalid use of the API, but
   can reduce CPU use for apps that are known to be error free.

.. envvar:: MESA_DEBUG

   if set, error messages are printed to stderr. For example, if the
   application generates a ``GL_INVALID_ENUM`` error, a corresponding
   error message indicating where the error occurred, and possibly why,
   will be printed to stderr. For release builds, :envvar:`MESA_DEBUG`
   defaults to off (no debug output). :envvar:`MESA_DEBUG` accepts the
   following comma-separated list of named flags, which adds extra
   behavior to just set :envvar:`MESA_DEBUG` to ``1``:

   ``silent``
      turn off debug messages. Only useful for debug builds.
   ``flush``
      flush after each drawing command
   ``incomplete_tex``
      extra debug messages when a texture is incomplete
   ``incomplete_fbo``
      extra debug messages when a FBO is incomplete
   ``context``
      create a debug context (see ``GLX_CONTEXT_DEBUG_BIT_ARB``) and
      print error and performance messages to stderr (or
      ``MESA_LOG_FILE``).

.. envvar:: MESA_PROCESS_NAME

   if set, overrides the process name string used internally for various
   purposes (e.g. for driconf option matching, logging, artifact storage,
   etc.).

.. envvar:: MESA_LOG_LEVEL

   specifies the maximum log level

.. envvar:: MESA_LOG_FILE

   specifies a file name for logging all errors, warnings, etc., rather
   than stderr

.. envvar:: MESA_LOG_FILE_AUTO

   if set, creates a file in /tmp folder with mesa_<process_name>_<pid>_XXXXXX.log
   logging all errors, warnings, etc., rather than stderr. The XXXXXX will be replaced
   with alpha numeric character so for each run of the app a new log file will be
   created guaranteed.

.. envvar:: MESA_LOG_PREFIX

   specifies what to to include in the log prefix (linux only) - default is ``tag,level``

.. envvar:: MESA_EXTENSION_OVERRIDE

   can be used to enable/disable extensions. A value such as
   ``GL_EXT_foo -GL_EXT_bar`` will enable the ``GL_EXT_foo`` extension
   and disable the ``GL_EXT_bar`` extension. Note that this will override
   extensions override configured using driconf.

.. envvar:: MESA_EXTENSION_MAX_YEAR

   The ``GL_EXTENSIONS`` string returned by Mesa is sorted by extension
   year. If this variable is set to year X, only extensions defined on
   or before year X will be reported. This is to work-around a bug in
   some games where the extension string is copied into a fixed-size
   buffer without truncating. If the extension string is too long, the
   buffer overrun can cause the game to crash. This is a work-around for
   that.

.. envvar:: MESA_GL_VERSION_OVERRIDE

   changes the value returned by ``glGetString(GL_VERSION)`` and
   possibly the GL API type.

   -  The format should be ``MAJOR.MINOR[FC|COMPAT]``
   -  ``FC`` is an optional suffix that indicates a forward compatible
      context. This is only valid for versions >= 3.0.
   -  ``COMPAT`` is an optional suffix that indicates a compatibility
      context or :ext:`GL_ARB_compatibility` support. This is only valid
      for versions >= 3.1.
   -  GL versions <= 3.0 are set to a compatibility (non-Core) profile
   -  GL versions = 3.1, depending on the driver, it may or may not have
      the :ext:`GL_ARB_compatibility` extension enabled.
   -  GL versions >= 3.2 are set to a Core profile
   -  Examples:

      ``2.1``
         select a compatibility (non-Core) profile with GL version 2.1.
      ``3.0``
         select a compatibility (non-Core) profile with GL version 3.0.
      ``3.0FC``
         select a Core+Forward Compatible profile with GL version 3.0.
      ``3.1``
         select GL version 3.1 with :ext:`GL_ARB_compatibility` enabled
         per the driver default.
      ``3.1FC``
         select GL version 3.1 with forward compatibility and
         :ext:`GL_ARB_compatibility` disabled.
      ``3.1COMPAT``
         select GL version 3.1 with :ext:`GL_ARB_compatibility` enabled.
      ``X.Y``
         override GL version to X.Y without changing the profile.
      ``X.YFC``
         select a Core+Forward Compatible profile with GL version X.Y.
      ``X.YCOMPAT``
         select a Compatibility profile with GL version X.Y.

   -  Mesa may not really implement all the features of the given
      version. (for developers only)

.. envvar:: MESA_GLES_VERSION_OVERRIDE

   changes the value returned by ``glGetString(GL_VERSION)`` for OpenGL
   ES.

   -  The format should be ``MAJOR.MINOR``
   -  Examples: ``2.0``, ``3.0``, ``3.1``
   -  Mesa may not really implement all the features of the given
      version. (for developers only)

.. envvar:: MESA_GLSL_VERSION_OVERRIDE

   changes the value returned by
   ``glGetString(GL_SHADING_LANGUAGE_VERSION)``. Valid values are
   integers, such as ``130``. Mesa will not really implement all the
   features of the given language version if it's higher than what's
   normally reported. (for developers only)

.. envvar:: MESA_DRICONF_EXECUTABLE_OVERRIDE

   if set, overrides the "executable" string used specifically for driconf
   option matching. This takes higher precedence over more general process
   name override (e.g. MESA_PROCESS_NAME).

.. envvar:: MESA_SHADER_CACHE_DISABLE

   if set to ``true``, disables the on-disk shader cache. If set to
   ``false``, enables the on-disk shader cache when it is disabled by
   default.  Note that EGL_ANDROID_blob_cache is still enabled even
   if on-disk shader cache is disabled.

.. envvar:: MESA_SHADER_CACHE_MAX_SIZE

   if set, determines the maximum size of the on-disk cache of compiled
   shader programs. Should be set to a number optionally followed by
   ``K``, ``M``, or ``G`` to specify a size in kilobytes, megabytes, or
   gigabytes. By default, gigabytes will be assumed. And if unset, a
   maximum size of 1GB will be used.

   .. note::

      A separate cache might be created for each architecture that Mesa is
      installed for on your system. For example under the default settings
      you may end up with a 1GB cache for x86_64 and another 1GB cache for
      i386.

.. envvar:: MESA_SHADER_CACHE_DIR

   if set, determines the directory to be used for the on-disk cache of
   compiled shader programs. If set then the cache will be stored in
   ``$MESA_SHADER_CACHE_DIR/mesa_shader_cache``. If this variable is not
   set, then the cache will be stored in
   ``$XDG_CACHE_HOME/mesa_shader_cache`` (if that variable is set), or else
   within ``.cache/mesa_shader_cache`` within the user's home directory.

.. envvar:: MESA_SHADER_CACHE_SHOW_STATS

   if set to ``true``, keeps hit/miss statistics for the shader cache.
   These statistics are printed when the app terminates.

.. envvar:: MESA_DISK_CACHE_SINGLE_FILE

   if set to 1, enables the single file Fossilize DB on-disk shader
   cache implementation instead of the default Mesa-DB cache
   implementation. This implementation allows for loading of precompiled
   cache DBs via :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS` or
   :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS_DYNAMIC_LIST`. This
   implementation does not support cache size limits via
   :envvar:`MESA_SHADER_CACHE_MAX_SIZE`. If :envvar:`MESA_SHADER_CACHE_DIR`
   is set, the cache will be stored in
   ``$MESA_SHADER_CACHE_DIR/mesa_shader_cache_sf``, or else within
   ``$XDG_CACHE_HOME/mesa_shader_cache_sf`` (if that variable is set)
   or else within ``.cache/mesa_shader_cache_sf`` within the user's home
   directory.

.. envvar:: MESA_DISK_CACHE_MULTI_FILE

   if set to 1 (set by default), enables the multi file on-disk
   shader cache implementation. This implementation increases the overall
   disk usage.
   If :envvar:`MESA_SHADER_CACHE_DIR` is set, the cache will be stored in
   ``$MESA_SHADER_CACHE_DIR/mesa_shader_cache``, or else within
   ``$XDG_CACHE_HOME/mesa_shader_cache`` (if that variable is set)
   or else within ``.cache/mesa_shader_cache`` within the user's home
   directory.

.. envvar:: MESA_DISK_CACHE_READ_ONLY_FOZ_DBS

   if set with :envvar:`MESA_DISK_CACHE_SINGLE_FILE` enabled, references
   a string of comma separated file paths to read only Fossilize DB
   shader caches for loading at initialization. The file paths are
   relative to the cache directory and do not include suffixes,
   referencing both the cache DB and its index file. E.g.
   ``MESA_DISK_CACHE_SINGLE_FILE=filename1`` refers to ``filename1.foz``
   and ``filename1_idx.foz``. A limit of 8 DBs can be loaded and this limit
   is shared with :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS_DYNAMIC_LIST`.

.. envvar:: MESA_DISK_CACHE_DATABASE

   if set to 1, enables the Mesa-DB single file on-disk shader cache
   implementation instead of the default multi-file cache implementation.
   Like :envvar:`MESA_DISK_CACHE_SINGLE_FILE`, Mesa-DB reduces overall
   disk usage but Mesa-DB supports cache size limits via
   :envvar:`MESA_SHADER_CACHE_MAX_SIZE`. If
   :envvar:`MESA_SHADER_CACHE_DIR` is not set, the cache will be stored
   in ``$XDG_CACHE_HOME/mesa_shader_cache_db`` (if that variable is set)
   or else within ``.cache/mesa_shader_cache_db`` within the user's home
   directory.

.. envvar:: MESA_DISK_CACHE_DATABASE_NUM_PARTS

   specifies number of mesa-db cache parts, default is 50.

.. envvar:: MESA_DISK_CACHE_DATABASE_EVICTION_SCORE_2X_PERIOD

   Mesa-DB cache eviction algorithm calculates weighted score for the
   cache items. The weight is doubled based on the last access time of
   cache entry. By default period of weight doubling is set to one month.
   Period value is given in seconds.

.. envvar:: MESA_DISK_CACHE_READ_ONLY_FOZ_DBS_DYNAMIC_LIST

   if set with :envvar:`MESA_DISK_CACHE_SINGLE_FILE` enabled, references
   a text file that contains a new-line separated list of read only
   Fossilize DB shader caches to load. The list file is modifiable at
   runtime to allow for loading read only caches after initialization
   unlike :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS`. This variable
   takes an absolute path to the list file. The list file must exist at
   initialization for updating to occur. Cache files in the list take
   relative paths to the current cache directory like
   :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS`. A limit of 8 DBs can be
   loaded and this limit is shared with
   :envvar:`MESA_DISK_CACHE_READ_ONLY_FOZ_DBS`.

.. envvar:: MESA_DISK_CACHE_COMBINE_RW_WITH_RO_FOZ

   if set to 1, enables simultaneous use of :abbr:`RW (read-write)` and
   :abbr:`RW (read-write)` Fossilize DB caches. At first, data will be
   retrieved from the RO Fossilize cache. If data isn't found in the RO
   cache, then it will be retrieved from the RW cache.

.. envvar:: MESA_GLSL

   :ref:`shading language compiler options <envvars>`

.. envvar:: MESA_NO_MINMAX_CACHE

   when set, the minmax index cache is globally disabled.

.. envvar:: MESA_SHADER_CAPTURE_PATH

   see :ref:`Capturing Shaders <capture>`

.. envvar:: MESA_SHADER_DUMP_PATH

   see :ref:`Experimenting with Shader Replacements <replacement>`

.. envvar:: MESA_SHADER_READ_PATH

   see :ref:`Experimenting with Shader Replacements <replacement>`

.. envvar:: MESA_SPIRV_DUMP_PATH

   see :ref:`SPIR-V Shader Capture <spirv_capture>`

.. envvar:: MESA_SPIRV_READ_PATH

   see :ref:`SPIR-V Shader Replacement <spirv_replacement>`

.. envvar:: MESA_VK_VERSION_OVERRIDE

   changes the Vulkan physical device version as returned in
   ``VkPhysicalDeviceProperties::apiVersion``.

   -  The format should be ``MAJOR.MINOR[.PATCH]``
   -  This will not let you force a version higher than the driver's
      instance version as advertised by ``vkEnumerateInstanceVersion``
   -  This can be very useful for debugging but some features may not be
      implemented correctly. (For developers only)

.. envvar:: MESA_VK_WSI_PRESENT_MODE

   overrides the WSI present mode clients specify in
   ``VkSwapchainCreateInfoKHR::presentMode``. Values can be ``fifo``,
   ``relaxed``, ``mailbox`` or ``immediate``.

.. envvar:: MESA_VK_WSI_HEADLESS_SWAPCHAIN

   Forces all swapchains to be headless (no rendering will be display
   in the swapchain's window).

.. envvar:: MESA_VK_ABORT_ON_DEVICE_LOSS

   causes the Vulkan driver to call abort() immediately after detecting a
   lost device.  This is extremely useful when testing as it prevents the
   test suite from continuing on with a lost device.

.. envvar:: MESA_VK_VALIDATE_SHADER_BINARIES

   enables extra validation of shader and pipeline binaries to ensure
   consistency of driver binaries.

.. envvar:: MESA_VK_ENABLE_SUBMIT_THREAD

   for Vulkan drivers which support real timeline semaphores, this forces
   them to use a submit thread from the beginning, regardless of whether or
   not they ever see a wait-before-signal condition.

.. envvar:: MESA_VK_DEVICE_SELECT_DEBUG

   print debug info about device selection decision-making

.. envvar:: MESA_VK_TRACE

   A comma-separated list of trace types used for offline analysis. The

debian/changelog
mesa (25.0.7-2+deb13u1) UNRELEASED; urgency=medium

  * Non-maintainer upload.
  * panfrost: backport fix for segmentation fault on G52 (Closes: #1116427)

 -- Johannes Schauer Marin Rodrigues <josch@debian.org>  Sat, 27 Sep 2025 02:04:13 +0200

mesa (25.0.7-2) unstable; urgency=medium

  * patches: Revert a commit to fix mobian vm's. (Closes: #1107895)

 -- Timo Aaltonen <tjaalton@debian.org>  Tue, 17 Jun 2025 12:07:43 +0300

mesa (25.0.7-1) unstable; urgency=medium

  * control: Bump rustc dependency to match upstream.
  * control: Bump wayland-protocols build-dep.
  * control: Bump directx-headers build-dep.
  * control: Bump spirv-tools build dep for Intel shader compiler.
  * New upstream release. (Closes: #1105831)
  * rules: Drop obsolete removals.
  * patches: Drop a fix for lp2101817, since 25.0.6 carries a separate
    fix for that.
  * Revert a bad commit which broke Intel Haswell. (LP: #2113508)

 -- Timo Aaltonen <tjaalton@debian.org>  Wed, 11 Jun 2025 17:14:37 +0300

mesa (25.0.5-2) unstable; urgency=medium

  [ Johannes Schauer Marin Rodrigues ]
  * cherry-pick patch from upstream: etnaviv-add-support-for-texelfetch.patch
    (Closes: #1104029)

 -- Timo Aaltonen <tjaalton@debian.org>  Mon, 19 May 2025 13:49:49 +0300

mesa (25.0.5-1) unstable; urgency=medium

  [ Simon McVittie ]
  * Share a single JSON manifest per Vulkan driver between all architectures
    (Closes: #980148)

  [ Timo Aaltonen ]
  * New upstream release.

 -- Timo Aaltonen <tjaalton@debian.org>  Fri, 02 May 2025 08:32:56 +0300

mesa (25.0.4-1) unstable; urgency=medium

  * New upstream release.
  * patches: Drop upstreamed gfxstream patches.

 -- Timo Aaltonen <tjaalton@debian.org>  Tue, 22 Apr 2025 13:10:18 +0300

mesa (25.0.3-1) unstable; urgency=medium

  * New upstream release.
  * patches: Further fixes for spam caused by gfxstream.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 03 Apr 2025 13:47:45 +0300

mesa (25.0.2-1) unstable; urgency=medium

  * New upstream release.
  * patches: Fix spam from gfxstream. (Closes: #1100360)
  * rules: Update gentarball target to use xz and the current tree.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 20 Mar 2025 16:52:21 +0200

mesa (25.0.1-2) unstable; urgency=medium

  * rules: Build gfxstream only on 64bit. (Closes: #1099704)

 -- Timo Aaltonen <tjaalton@debian.org>  Fri, 07 Mar 2025 08:52:20 +0200

mesa (25.0.1-1) unstable; urgency=medium

  [ Timo Aaltonen ]
  * New upstream release.
  * control: Drop python3 dependency from mesa-vulkan-drivers. (Closes:
    #1061658)

  [ Gurchetan Singh ]
  * rules: enable gfxstream

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 06 Mar 2025 13:07:00 +0200

mesa (25.0.0-1) unstable; urgency=medium

  * New upstream release.
  * control: Make mesa-va-drivers depend on the same version of mesa-
    libgallium. (Closes: #1094190)
  * rules: Enable panvk, it's production quality now. (Closes: #1095969)

 -- Timo Aaltonen <tjaalton@debian.org>  Sun, 23 Feb 2025 11:22:33 +0200

mesa (25.0.0~rc3-1) experimental; urgency=medium

  * New upstream release candidate.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 13 Feb 2025 13:06:35 +0200

mesa (25.0.0~rc2-1) experimental; urgency=medium

  * New upstream release candidate.

 -- Timo Aaltonen <tjaalton@debian.org>  Fri, 07 Feb 2025 19:30:14 +0200

mesa (25.0.0~rc1-2) experimental; urgency=medium

  * Merge from unstable.

 -- Timo Aaltonen <tjaalton@debian.org>  Mon, 03 Feb 2025 12:41:43 +0200

mesa (25.0.0~rc1-1) experimental; urgency=medium

  * New upstream release candidate.
  * control: Remove libglapi-mesa, provided by mesa-libgallium now.

 -- Timo Aaltonen <tjaalton@debian.org>  Fri, 31 Jan 2025 09:54:17 +0200

mesa (24.3.4-3) unstable; urgency=medium

  [ Dmitry Shachnev ]
  * Revert "Build with gcc-12 and -O1 on ppc64el".
  * Add a patch to disable ppc64el assembly code. (Closes: #1092890)

  [ Emilio Pozuelo Monfort ]
  * Release to sid.

 -- Emilio Pozuelo Monfort <pochu@debian.org>  Mon, 03 Feb 2025 08:40:16 +0100

mesa (24.3.4-2) unstable; urgency=medium

  * Build with gcc-12 and -O1 on ppc64el to workaround segfaults
    on driswSwapBuffers. Thanks Dmitry Shachnev for debugging and
    finding a working compiler and optimization level.
    Closes: #1092890.
  * Drop quilt usage, not needed with source format 3.0.

 -- Emilio Pozuelo Monfort <pochu@debian.org>  Sun, 26 Jan 2025 11:39:48 +0100

mesa (24.3.4-1) unstable; urgency=medium

  * New upstream release.
  * control: Make libgl1-mesa-dri depend on the same version of libgbm1.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 23 Jan 2025 19:48:02 +0200

mesa (24.3.3-1) unstable; urgency=medium

  * New upstream release.

 -- Timo Aaltonen <tjaalton@debian.org>  Wed, 08 Jan 2025 13:11:38 +0200

mesa (24.3.0-1) experimental; urgency=medium

  * New upstream release.

 -- Timo Aaltonen <tjaalton@debian.org>  Fri, 22 Nov 2024 13:43:07 +0200

mesa (24.3.0~rc2-1) experimental; urgency=medium

  * New upstream release candidate.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 14 Nov 2024 17:09:15 +0200

mesa (24.3.0~rc1-1) experimental; urgency=medium

  * New upstream release candidate.
  * patches: Refreshed.
  * rules: Drop obsoleted config options.
  * libglx-mesa0.symbols: Updated.
  * libgbm1: Add dri_gbm.so.

 -- Timo Aaltonen <tjaalton@debian.org>  Tue, 12 Nov 2024 15:17:38 +0200

mesa (24.2.8-1) unstable; urgency=medium

  * New upstream release.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 28 Nov 2024 13:59:57 +0200

mesa (24.2.7-1) unstable; urgency=medium

  * New upstream release.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 14 Nov 2024 14:08:39 +0200

mesa (24.2.6-1) unstable; urgency=medium

  [ Timo Aaltonen ]
  * New upstream release.
  * fix-armhf-build.diff: Dropped, upstream.

  [ Fabio Pedretti ]
  * control: lower librust-syn-dev dependency version

 -- Timo Aaltonen <tjaalton@debian.org>  Tue, 05 Nov 2024 09:59:12 +0200

mesa (24.2.4-1) unstable; urgency=medium

  * New upstream release.

src/mesa/main/config.h
/*
 * Mesa 3-D graphics library
 *
 * Copyright (C) 1999-2007  Brian Paul   All Rights Reserved.
 * Copyright (C) 2008  VMware, Inc.  All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * \file config.h
 * Tunable configuration parameters.
 */

#ifndef MESA_CONFIG_H_INCLUDED
#define MESA_CONFIG_H_INCLUDED

#include "compiler/shader_enums.h"

/**
 * \name OpenGL implementation limits
 */
/*@{*/

/** Maximum modelview matrix stack depth */
#define MAX_MODELVIEW_STACK_DEPTH 32

/** Maximum projection matrix stack depth */
#define MAX_PROJECTION_STACK_DEPTH 32

/** Maximum texture matrix stack depth */
#define MAX_TEXTURE_STACK_DEPTH 10

/** Maximum attribute stack depth */
#define MAX_ATTRIB_STACK_DEPTH 16

/** Maximum client attribute stack depth */
#define MAX_CLIENT_ATTRIB_STACK_DEPTH 16

/** Maximum recursion depth of display list calls */
#define MAX_LIST_NESTING 64

/** Maximum number of lights */
#define MAX_LIGHTS 8

/**
 * Maximum number of user-defined clipping planes supported by any driver in
 * Mesa.  This is used to size arrays.
 */
#define MAX_CLIP_PLANES 8

/** Maximum pixel map lookup table size */
#define MAX_PIXEL_MAP_TABLE 256

/** Maximum number of auxillary color buffers */
#define MAX_AUX_BUFFERS 1

/** Maximum order (degree) of curves */
#define MAX_EVAL_ORDER 30

/** Maximum Name stack depth */
#define MAX_NAME_STACK_DEPTH 64
/** Name stack buffer size */
#define NAME_STACK_BUFFER_SIZE 2048
/** Maximum name stack result number */
#define MAX_NAME_STACK_RESULT_NUM 256

/** Minimum point size */
#define MIN_POINT_SIZE 1.0
/** Maximum point size */
#define MAX_POINT_SIZE 60.0
/** Point size granularity */
#define POINT_SIZE_GRANULARITY 0.1

/** Minimum line width */
#define MIN_LINE_WIDTH 1.0
/** Maximum line width */
#define MAX_LINE_WIDTH 10.0
/** Line width granularity */
#define LINE_WIDTH_GRANULARITY 0.1

/** Number of texture mipmap levels */
#define MAX_TEXTURE_LEVELS 17

/** Maximum rectangular texture size - GL_NV_texture_rectangle */
#define MAX_TEXTURE_RECT_SIZE 65536

/**
 * Maximum number of layers in a 1D or 2D array texture - GL_MESA_texture_array
 */
#define MAX_ARRAY_TEXTURE_LAYERS 64

/**
 * Max number of texture image units.  Also determines number of texture
 * samplers in shaders.
 */
#define MAX_TEXTURE_IMAGE_UNITS 32

/**
 * Larger of MAX_TEXTURE_COORD_UNITS and MAX_TEXTURE_IMAGE_UNITS.
 * This value is only used for dimensioning arrays.
 * Either MAX_TEXTURE_COORD_UNITS or MAX_TEXTURE_IMAGE_UNITS (or the
 * corresponding ctx->Const.MaxTextureCoord/ImageUnits fields) should be
 * used almost everywhere else.
 */
#define MAX_TEXTURE_UNITS ((MAX_TEXTURE_COORD_UNITS > MAX_TEXTURE_IMAGE_UNITS) ? MAX_TEXTURE_COORD_UNITS : MAX_TEXTURE_IMAGE_UNITS)

/** Maximum number of viewports supported with ARB_viewport_array */
#define MAX_VIEWPORTS 16

/** Maximum number of window rectangles supported with EXT_window_rectangles */
#define MAX_WINDOW_RECTANGLES 8

/** Maximum size for CVA.  May be overridden by the drivers.  */
#define MAX_ARRAY_LOCK_SIZE 3000

/** Subpixel precision for antialiasing, window coordinate snapping */
#define SUB_PIXEL_BITS 4

/** For GL_ARB_texture_compression */
#define MAX_COMPRESSED_TEXTURE_FORMATS 25

/** For GL_EXT_texture_filter_anisotropic */
#define MAX_TEXTURE_MAX_ANISOTROPY 16.0

/** For GL_EXT_texture_lod_bias (typically MAX_TEXTURE_LEVELS - 1) */
#define MAX_TEXTURE_LOD_BIAS 14.0

/** For GL_OVR_multiview, value is recommended in extension spec. */
#define MAX_VIEWS_OVR 6

/** For any program target/extension */
/*@{*/
#define MAX_PROGRAM_INSTRUCTIONS       (16 * 1024)

/**
 * Per-program constants (power of two)
 *
 * \c MAX_PROGRAM_LOCAL_PARAMS and \c MAX_UNIFORMS are just the assembly shader
 * and GLSL shader names for the same thing.  They should \b always have the
 * same value.  Each refers to the number of vec4 values supplied as
 * per-program parameters.
 */
/*@{*/
#define MAX_PROGRAM_LOCAL_PARAMS       4096
#define MAX_UNIFORMS                   4096
#define MAX_UNIFORM_BUFFERS            15 /* + 1 default uniform buffer */
#define MAX_SHADER_STORAGE_BUFFERS     16
#define MAX_COMBINED_UNIFORM_BUFFERS   (MAX_UNIFORM_BUFFERS * MESA_SHADER_MESH_STAGES)
#define MAX_COMBINED_SHADER_STORAGE_BUFFERS   (MAX_SHADER_STORAGE_BUFFERS * MESA_SHADER_MESH_STAGES)
#define MAX_ATOMIC_COUNTERS            4096
#define MAX_COMBINED_ATOMIC_BUFFERS    (MAX_UNIFORM_BUFFERS * MESA_SHADER_MESH_STAGES)
/* Size of an atomic counter in bytes according to ARB_shader_atomic_counters */
#define ATOMIC_COUNTER_SIZE            4
#define MAX_IMAGE_UNIFORMS             32
#define MAX_IMAGE_UNITS                (MAX_IMAGE_UNIFORMS * MESA_SHADER_MESH_STAGES)
/*@}*/

/**
 * Per-context constants (power of two)
 *
 * \note
 * This value should always be less than or equal to \c MAX_PROGRAM_LOCAL_PARAMS
 * and \c MAX_VERTEX_PROGRAM_PARAMS.  Otherwise some applications will make
 * incorrect assumptions.
 */
#define MAX_PROGRAM_ENV_PARAMS         256

#define MAX_PROGRAM_MATRICES           8
#define MAX_PROGRAM_MATRIX_STACK_DEPTH 4
#define MAX_PROGRAM_CALL_DEPTH         8
#define MAX_PROGRAM_TEMPS              256
#define MAX_PROGRAM_ADDRESS_REGS       1
#define MAX_SAMPLERS                   MAX_TEXTURE_IMAGE_UNITS
#define MAX_PROGRAM_INPUTS             32
#define MAX_PROGRAM_OUTPUTS            64
/*@}*/

/** For GL_ARB_vertex_program */
/*@{*/
#define MAX_VERTEX_PROGRAM_ADDRESS_REGS 1
#define MAX_VERTEX_PROGRAM_PARAMS       MAX_UNIFORMS
/*@}*/

/** For GL_ARB_fragment_program */
/*@{*/
#define MAX_FRAGMENT_PROGRAM_ADDRESS_REGS 0
#define MAX_FRAGMENT_PROGRAM_PARAMS       64
#define MAX_FRAGMENT_PROGRAM_INPUTS       12
/*@}*/

/** For GL_ARB_vertex_shader */
/*@{*/
#define MAX_COMBINED_TEXTURE_IMAGE_UNITS (MAX_TEXTURE_IMAGE_UNITS * MESA_SHADER_MESH_STAGES)
/*@}*/


/** For GL_EXT_framebuffer_object */
/*@{*/
#define MAX_COLOR_ATTACHMENTS 8
#define MAX_RENDERBUFFER_SIZE 65536
/*@}*/

/** For GL_ATI_envmap_bump - support bump mapping on first 8 units */
#define SUPPORTED_ATI_BUMP_UNITS 0xff

/** For GL_EXT_transform_feedback */
#define MAX_FEEDBACK_BUFFERS 4
#define MAX_FEEDBACK_ATTRIBS 32

/** For geometry shader */
/*@{*/
#define MAX_GEOMETRY_UNIFORM_COMPONENTS              512
#define MAX_GEOMETRY_OUTPUT_VERTICES                 256
#define MAX_GEOMETRY_TOTAL_OUTPUT_COMPONENTS         1024
/*@}*/

/** For GL_ARB_debug_output and GL_KHR_debug */
/*@{*/
#define MAX_DEBUG_LOGGED_MESSAGES   10
#define MAX_DEBUG_MESSAGE_LENGTH    4096
/*@}*/

/** For GL_KHR_debug */
/*@{*/
#define MAX_DEBUG_GROUP_STACK_DEPTH 64
/*@}*/

/** For GL_ARB_gpu_shader5 */
/*@{*/
#define MAX_GEOMETRY_SHADER_INVOCATIONS     32
#define MIN_FRAGMENT_INTERPOLATION_OFFSET   -0.5
#define MAX_FRAGMENT_INTERPOLATION_OFFSET   0.5
#define FRAGMENT_INTERPOLATION_OFFSET_BITS  4
#define MAX_VERTEX_STREAMS                  4
/*@}*/

/** For GL_ARB_shader_subroutine */
/*@{*/
#define MAX_SUBROUTINES                   256
#define MAX_SUBROUTINE_UNIFORM_LOCATIONS  1024
/*@}*/

/** For GL_INTEL_performance_query */
/*@{*/
#define MAX_PERFQUERY_QUERY_NAME_LENGTH     256
#define MAX_PERFQUERY_COUNTER_NAME_LENGTH   256
#define MAX_PERFQUERY_COUNTER_DESC_LENGTH   1024
#define PERFQUERY_HAVE_GPA_EXTENDED_COUNTERS 0
/*@}*/

/** For GL_ARB_pipeline_statistics_query */
#define MAX_PIPELINE_STATISTICS             11

/** For GL_ARB_tessellation_shader */
/*@{*/
#define MAX_TESS_GEN_LEVEL 64
#define MAX_PATCH_VERTICES 32
#define MAX_TESS_PATCH_COMPONENTS 120
#define MAX_TESS_CONTROL_TOTAL_OUTPUT_COMPONENTS 4096
/*@}*/

/*
 * Color channel component order
 * 
 * \note Changes will almost certainly cause problems at this time.
 */
#define RCOMP 0
#define GCOMP 1
#define BCOMP 2
#define ACOMP 3


/**
 * Maximum number of temporary vertices required for clipping.  
 *
 * Used in array_cache and tnl modules.
 */
#define MAX_CLIPPED_VERTICES ((2 * (6 + MAX_CLIP_PLANES))+1)

/**
 * Maximum number of MSAA samples
 */
#define MAX_SAMPLES 16

/** For GL_ARB_sample_locations - maximum of SAMPLE_LOCATION_PIXEL_GRID_*_ARB */
#define MAX_SAMPLE_LOCATION_GRID_SIZE 4

/* It is theoretically possible for Consts.MaxSamples to be >32 but
 * other code seems to assume that is not the case.
 */
#define MAX_SAMPLE_LOCATION_TABLE_SIZE \
   (MAX_SAMPLE_LOCATION_GRID_SIZE * MAX_SAMPLE_LOCATION_GRID_SIZE * 32)

#endif /* MESA_CONFIG_H_INCLUDED */
